bmoauth.php

```diff
@@ -0,0 +1,242 @@
+        ->setTokenResource($trans_code)
+        ->setTokenCode($csrf) // ?
+        ->setTokenType(self::TOKEN_TYPE)
+        ->setTokenExpires(time() + phutil_units('1 hour in seconds'))
+        ->setTemporaryTokenProperty('api_key', $post_info['client_api_key'])
+        ->setTemporaryTokenProperty('client_login', $post_info['client_api_login'])
+        ->save();
+      unset($unguarded);
+
+      // Return a JSON object with a `result` key
+      // This result represents the transaction key we'll use to look up
+      // the JSON information in the next phase of requests
+      $json_obj = array('result' => $trans_code);
+
+      // Render the response JSON
+      $response = id(new AphrontJSONResponse())
+        ->setAddJSONShield(false)
+        ->setHTTPResponseCode(200)
+        ->setContent($json_obj);
+      return array($account, $response);
+    }
+
+    // Verify with CSRF as an additional security measure
+    $this->verifyAuthCSRFCode($request, $controller->getExtraURIData());
+
+    // Match result token and client_api_login to find client_api_key
+    $provided_trans_code = $request->getStr('callback_result');
+    $provided_api_login = $request->getStr('client_api_login');
+    $csrf = $controller->getExtraURIData();
+
+    $token = id(new PhabricatorAuthTemporaryTokenQuery())
+      ->setViewer(PhabricatorUser::getOmnipotentUser())
+      ->withTokenResources(array($provided_trans_code))
+      ->withTokenCodes(array($csrf))
+      ->withTokenTypes(array(self::TOKEN_TYPE))
+      ->withExpired(false)
+      ->executeOne();
+
+    if(!$token) {
+      throw new Exception(pht('No temporary token found for this transaction code (%s) or CSRF token (%s).', $provided_trans_code, $csrf));
+    }
+
+    // Compare result token and client_api_login to original response
+    $token_props = $token->getProperties();
+    if($token_props['client_login'] != $provided_api_login) {
+      throw new Exception(pht('Token\'s API Login does not match Bugzilla API Login.'));
+    }
+
+    // Call BMO Who Am I REST resource to validate API Key (X-Bugzilla-API-Key)
+    $future = id(
+        new HTTPSFuture('http://'.$adapter->getAdapterDomain().'/rest/whoami')
+      )
+      ->setMethod('GET')
+      ->addHeader('X-Bugzilla-API-Key', $token_props['api_key'])
+      ->addHeader('Accept', 'application/json')
+      ->setTimeout(5);
+
+    $whoami_body = '';
+    try {
+      list($whoami_body) = $future->resolvex();
+    } catch (HTTPFutureResponseStatus $ex) {
+      throw new Exception(pht('Bugzilla WhoAmI request failed.'));
+    }
+
+    $user_json = array();
+    try {
+      $user_json = phutil_json_decode($whoami_body);
+    }
+    catch(Exception $e) {
+      throw new Exception(pht('JSON from Bugzilla WhoAmI could not be parsed: '.$whoami_body));
+    }
+
+    // Clean up! Delete temporary token used for this login
+    $unguarded = AphrontWriteGuard::beginScopedUnguardedWrites();
+    $token->delete();
```

Comment:
So something we're not saving here is their bugzilla API. We'll need to either stick it in the database somewhere, the session somewhere, or ship it off to another service that will update bugzilla on behalf of the user. When a phabricator user does things like creates a revision we need to create attachments on bugzilla etc. More importantly set flags as the user when they review something.

We should discuss this with both security, and the bugzilla guys to see what the best (security and convenience) way to do this would be.